### PR TITLE
Add quickstart for Configuration API

### PR DESCRIPTION
**Pull request by sarsharma**
Signed-off-by: Sarthak Sharma <sartsharma@microsoft.com>

# Description

Adding a quickstart section for configuration building block.

## Issue reference

We strive to have all PRs being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_https://github.com/dapr/quickstarts/issues/554

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] The quickstart code compiles correctly
* [ ] You've tested new builds of the quickstart if you changed quickstart code
* [x] You've updated the quickstart's README if necessary
* [x] If you have changed the steps for a quickstart be sure that you have updated the automated validation accordingly. All of our quickstarts have annotations that allow them to be executed automatically as code. For more information see [mechanical-markdown](https://github.com/dapr/mechanical-markdown#readme). For user guide with examples see [Examples](https://github.com/dapr/mechanical-markdown/tree/main/examples#mechanical-markdown-by-example).

### CODE REVIEW

File: configuration/go/sdk/config-subscriber/app.go
Comment (akhilac1) [inline configuration/go/sdk/config-subscriber/app.go:18]: Any reason this variable name is in ALL CAPS? Is this intended to be a global variable? Please fix the case if otherwise.

File: configuration/go/sdk/config-subscriber/app.go
Comment (akhilac1) [inline configuration/go/sdk/config-subscriber/app.go:26]: please also print the error or users to report or troubleshoot in case of a error.
Will the code continue to execute in case of error also?

File: configuration/go/sdk/config-subscriber/app.go
Comment (akhilac1) [inline configuration/go/sdk/config-subscriber/app.go:46]: Why should the code panic in case of error in subscribing. Can we show the error and return?

File: configuration/go/sdk/config-subscriber/app.go
Comment (akhilac1) [inline configuration/go/sdk/config-subscriber/app.go:45]: Please also print the error here too

**Review by akhilac1 — CHANGES_REQUESTED** [review]
Thanks for putting this together @sarsharma. I left a few comments. Please take a look

File: configuration/go/sdk/config-subscriber/app.go
Comment (sarsharma) [reply configuration/go/sdk/config-subscriber/app.go:28]: App should exit incase it is unable to retrieve config items. Updated code to do the same.

File: configuration/go/sdk/config-subscriber/app.go
Comment (sarsharma) [reply configuration/go/sdk/config-subscriber/app.go]: updated to printing the error and then exiting sample app with non zero code

File: configuration/go/sdk/config-subscriber/app.go
Comment (sarsharma) [reply configuration/go/sdk/config-subscriber/app.go]: added

**Comment by amulyavarote** [comment]
Please add changes to the github workflows as well.

**Comment by amulyavarote** [comment]
Can we also maintain the consistency of having order-processor service in config API as well. We have order-processor and checkout service defined in all our examples. And we store orderId in all our samples. You can refer to state management examples. You can follow the similar samples for this as well.
https://github.com/dapr/quickstarts/blob/master/state_management/go/http/order-processor/app.go
https://github.com/dapr/quickstarts/blob/master/state_management/go/sdk/order-processor/app.go

File: configuration/go/http/config-subscriber/app.go
Comment (amulyavarote) [inline configuration/go/http/config-subscriber/app.go:1]: Change the folder name based on the existing samples.

File: configuration/go/http/config-subscriber/app.go
Comment (amulyavarote) [inline configuration/go/http/config-subscriber/app.go:20]: Use orderIds to maintain the consistency.

File: configuration/go/sdk/config-subscriber/app.go
Comment (amulyavarote) [inline configuration/go/sdk/config-subscriber/app.go:1]: Change the folder name based on the existing samples.

File: configuration/go/sdk/config-subscriber/app.go
Comment (amulyavarote) [inline configuration/go/sdk/config-subscriber/app.go:15]: Use orderIds to be consistent with other samples.

**Review by amulyavarote — CHANGES_REQUESTED** [review]
Let's try to maintain the consistency across all our samples(all the building blocks).

File: configuration/order-updater/main.go
Comment (shubham1172) [inline configuration/order-updater/main.go:12]: Is this supposed to be here? Or inside the go directory?

Comment (pravinpushkar) [reply configuration/order-updater/main.go:12]: @sarsharma I see why you have written this. Issue is people trying out lets say js quickstart might not have the setup for go. 

If it is only meant to be used by the pipeline to simulate the keys update behind the scene then should we  clearly mention in readmes that `order-updater` is only meant for simulating the behavior and optional.

Comment (shubham1172) [reply configuration/order-updater/main.go:12]: @pravinpushkar agreed, but building binaries for all OS/ARCH combinations might not be feasible. Even if you are keeping this go code here, I would recommend moving order-updater to `utils` or something. As @pravinpushkar mentioned, (a) it can be optional, and (b) you can provide REDIS CLI commands to do the same thing.

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:21]: can be removed.same below

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:36]: @dapr/maintainers-quickstarts should be 1.19.3?

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:37]: unsupported version. Change

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:38]: update

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:42]: ideally if possible change macOS.
Can be a separate issue to track the change also. cc @dapr/maintainers-quickstarts

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:51]: version should be updated.

File: .github/workflows/validate_new_quickstarts_configuration.yaml
Comment (mukundansundar) [inline .github/workflows/validate_new_quickstarts_configuration.yaml:53]: update

File: README.md
Comment (mukundansundar) [inline README.md:21]:
```suggestion
| [Configuration](./configuration) | Get configuration items as key/value pairs or subscribe to changes whenever a configuration item changes |
```

File: configuration/go/http/order-processor/app.go
Comment (mukundansundar) [inline configuration/go/http/order-processor/app.go:47]: does this need to be a go routine?

File: configuration/go/http/order-processor/app.go
Comment (mukundansundar) [inline configuration/go/http/order-processor/app.go:49]: properly shutdown server and go routines on exit.

File: configuration/go/http/order-processor/go.mod
Comment (mukundansundar) [inline configuration/go/http/order-processor/go.mod:3]: 1.19?

File: configuration/order-updater/main.go
Comment (mukundansundar) [reply configuration/order-updater/main.go:12]: I think calling it out as optional and providing a redis-cli command or docker based redis cli command to update config values is needed here.
We can make use of the `order-updater` only for validations(in README comments) and in go quickstart alone.

**Review by mukundansundar — COMMENTED** [review]
Should the `order-processor` be renamed as `config-processor` and `order-updater` be renamed as `config-updater`?

File: configuration/go/http/order-processor/app.go
Comment (sarsharma) [reply configuration/go/http/order-processor/app.go:47]: starting the http server blocks the main goroutine, hence running it on a separete goroutine

**Comment by sarsharma** [comment]
> Should the `order-processor` be renamed as `config-processor` and `order-updater` be renamed as `config-updater`?

@mukundansundar 
I had earlier kept it as `config-updater` and `config-subscriber`, but the folder names was changed to maintain consistency across all quickstarts. See the comment at the beginning by @amulyavarote

File: README.md
Comment (paulyuk) [inline README.md:20]: I never plan to have quickstart for observability or distributed lock.  Can i ask you to remove these Coming soons?  Actors ok to leave.

File: configuration/csharp/http/order-processor/Program.cs
Comment (paulyuk) [inline configuration/csharp/http/order-processor/Program.cs:13]: How common is it to pull config with multi keys?

File: configuration/csharp/http/order-processor/Program.csproj
Comment (paulyuk) [inline configuration/csharp/http/order-processor/Program.csproj:4]: at some point we'll do net7.0 but since it's early days i believe 6.0 is good.

File: configuration/csharp/sdk/README.md
Comment (paulyuk) [inline configuration/csharp/sdk/README.md:42]: String is clipped.  Does this pass tests?

File: configuration/go/http/order-processor/go.mod
Comment (paulyuk) [reply configuration/go/http/order-processor/go.mod:3]: all other quickstarts are 1.18.  Is there a need or benefit to forcing version higher?  @mukundansundar

File: configuration/order-updater/main.go
Comment (paulyuk) [reply configuration/order-updater/main.go:12]: I really worry about end users needing Go to try the other languages.

**Review by paulyuk — CHANGES_REQUESTED** [review]
Awesome work.  I love how this shows getting Configuration, and dynamic.  Great job on consistency.  

Please see comments.

Two top things to consider:
- is it more common for users to work with Configuration using a single key?  This quickstart biases to multiple keys, adding array complexity.  Is that what we want?
- do we really want all language samples to depend on GoLang/tools ?  I feel we should avoid that.

File: configuration/order-updater/main.go
Comment (sarsharma) [reply configuration/order-updater/main.go]: I removed `order-updater` and am now using `redis-cli` to add config items to the config store, and updating config items is optional for the user.

File: configuration/csharp/http/order-processor/Program.cs
Comment (sarsharma) [reply configuration/csharp/http/order-processor/Program.cs:13]: most SDKs have get_configuration() method that takes a list of keys as an input, so trying to show that through this example. Using a list will also show subscribing to multiple items

File: configuration/csharp/sdk/README.md
Comment (sarsharma) [reply configuration/csharp/sdk/README.md]: was using random generated id as value earlier, updated to keep it as a fixed value

**Comment by amulyavarote** [comment]
Hi! The PR looks great. Thanks for making the examples consistent. We have java examples missing in this PR. Do we have a separate PR for Java examples?

**Review by amulyavarote — DISMISSED** [review]

**Comment by amulyavarote** [comment]
I approved the existing examples for now.

File: configuration/go/http/order-processor/go.mod
Comment (mukundansundar) [reply configuration/go/http/order-processor/go.mod:3]: @paulyuk This is mainly to remain in step with all other Go repos in the Dapr project. 
Additionally we can make use of latest fixes and features in golang if any.

**Comment by sarsharma** [comment]
> Hi! The PR looks great. Thanks for making the examples consistent. We have java examples missing in this PR. Do we have a separate PR for Java examples?

Hi @amulyavarote, I was waiting for a new release of Java SDK, I have added examples for Java now.

**Comment by amulyavarote** [comment]
Let's wait for the workflows to succeed.

**Comment by amulyavarote** [comment]
Overall, the PR looks good! Thanks for the contribution here.

**Review by amulyavarote — DISMISSED** [review]

**Comment by sarsharma** [comment]
> Let's wait for the workflows to succeed.

@amulyavarote Validation passed on Ubuntu but is failing on MacOS, Looks like `mechanical-markdown` is having issues on MacOS 11 image, I am reverting MacOS runner image version for now, will open a new issue for this.

**Comment by mukundansundar** [comment]
@paulyuk Can you review this PR?

**Review by pravinpushkar — DISMISSED** [review]
lgtm. Workflow needs to be re-triggered.

**Comment by amulyavarote** [comment]
Changes look good to me. Just the workflows need to pass.

File: configuration/javascript/http/order-processor/index.js
Comment (greenie-msft) [inline configuration/javascript/http/order-processor/index.js:44]: Do we need sleeps? What exactly are we waiting for here? Dapr has health checks that we can use for this instead right?

**Comment by sarsharma** [comment]
Some workflows are timing out due to the unreliable nature of installing Docker on MacOS through the github action `docker-practice/actions-setup-docker@v1`. However, the tests themselves are valid since they are passing whenever the docker setup is successful over multiple runs. I have created an issue for this https://github.com/dapr/quickstarts/issues/755

File: configuration/javascript/http/order-processor/index.js
Comment (sarsharma) [reply configuration/javascript/http/order-processor/index.js:44]: Dapr waits for the app to be ready on the app-port. However, if I try to subscribe to config updates before the app is ready, I'll get an error since the app channel was not ready. App health check seems to be a preview feature and will require passing additional parameters to the CLI command, adding complexity to quickstarts.

**Comment by pravinpushkar** [comment]
> Some workflows are timing out due to the unreliable nature of installing Docker on MacOS through the github action `docker-practice/actions-setup-docker@v1`. This action is being used for all workflows. The tests themselves are valid since they are passing whenever the docker setup is successful over multiple runs. I have created an issue for this #755
> 
> Successful run for configuration api workflow https://github.com/dapr/quickstarts/actions/runs/3538439711

I think here also we can move to `Podman` in case of macos. See how it is done in `dapr/cli` today.

**Comment by mukundansundar** [comment]
Do we know as to why this is failing? https://github.com/dapr/quickstarts/actions/runs/3538439693/jobs/6107538083 @sarsharma ?

**Comment by sarsharma** [comment]
> Do we know as to why this is failing? https://github.com/dapr/quickstarts/actions/runs/3538439693/jobs/6107538083 @sarsharma ?

Looks like hello-kubernetes tutorial stdout is not matching in sequential order. 
Created an issue https://github.com/dapr/components-contrib/issues/2357, will take a look

File: configuration/csharp/http/README.md
Comment (paulyuk) [inline configuration/csharp/http/README.md:18]: I feel it would be a lot better for users and local mmd testing (`make validate`) if you could make this step instead do a `docker exec ...` of the command against the `dapr_redis` container that every Dapr CLI user already has.  

Could we please make this change?  then you will get the LGTM..

**Review by paulyuk — CHANGES_REQUESTED** [review]
Please see comment about dapr_redis container and docker exec.

File: configuration/csharp/http/README.md
Comment (sarsharma) [reply configuration/csharp/http/README.md]: Updated to use docker exec

**Review by paulyuk — APPROVED** [review]
LGTM.  Thanks a lot for this very nice contribution!

**Comment by paulyuk** [comment]
Validate Tutorials has been bad for a month and is likely from #737 and not this.  I'll merge.

**Pull request merged**
